### PR TITLE
feat: add Azure remote cache

### PR DESCRIPTION
**Pull request by arbreezy**
Closes # 

## 📑 Description

PR adds a new remote cache, integrating with Azure Blob storage offering.
It also includes the new protobuf schema.

Finally, it reworks the cache code path to make it easier to introduce more remote cache types in the future 

## ✅ Checks

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [x] I have updated the documentation as required
- [x] All the tests have passed

### CODE REVIEW

File: cmd/cache/add.go [hunk 1]
```diff
@@ -46,9 +51,15 @@ var addCmd = &cobra.Command{
...
-	addCmd.MarkFlagRequired("region")
-
+	addCmd.Flags().StringVarP(&region, "region", "r", "", "The region to use for the AWS S3 cache")
+	addCmd.Flags().StringVarP(&bucketname, "bucket", "b", "", "The name of the AWS S3 bucket to use for the cache")
+	addCmd.MarkFlagsRequiredTogether("region", "bucket")
+	addCmd.Flags().StringVarP(&storageAccount, "storageacc", "s", "", "The Azure storage account name of the container")
+	addCmd.Flags().StringVarP(&containerName, "container", "c", "", "The Azure container name to use for the cache")
+	addCmd.MarkFlagsRequiredTogether("storageacc", "container")
+	// Tedious check to ensure we don't include arguments from different providers
+	addCmd.MarkFlagsMutuallyExclusive("region", "storageacc")
```
Comment (AlexsJones) [inline cmd/cache/add.go:61]: I didnt know about this method!

**Review by AlexsJones — APPROVED** [review]
This has to be cautiously merged as it will break the caching functionality in the operator I believe @arbreezy

**Comment by AlexsJones** [comment]
Should this get merged in after you do the operator work or now @arbreezy ?

File: pkg/cache/cache.go [hunk 2]
```diff
@@ -7,6 +7,14 @@ import (
 	"github.com/spf13/viper"
 )
 
+type CacheType string
+
+const (
+	Azure     CacheType = "azure"
+	S3        CacheType = "s3"
+	FileBased CacheType = "file"
+)
```
Comment (matthisholleville) [inline pkg/cache/cache.go:16]: Shouldn't we handle cache providers in the same way we handle analyzers/filters?

```go
var remoteCacheProviderMap = map[string]ICache{
	"Azure":  &S3Cache{},
	"Azure": &AzureCache{},
        "File": &FileBasedCache{},
}

func ListRemoteCacheProviders() []string {
	remoteCacheProviderKeys := make([]string, 0, len(remoteCacheProviderMap))
	for k := range remoteCacheProviderMap {
		remoteCacheProviderKeys = append(remoteCacheProviderKeys, k)
	}
	return remoteCacheProviderKeys
}
```

**Review by matthisholleville — COMMENTED** [review]
Very good MR ! Nice job Aris

File: pkg/cache/cache.go (hunk 2 above)
Comment (arbreezy) [reply pkg/cache/cache.go:16]: a) I think there is not a big value out of having this as an interface, b) the list functionality in caches is listing files from your already configured cache backend and not a list of supported caches from.

**Review by AlexsJones — APPROVED** [review]

**Pull request merged**